Source/CardStackController.m

```diff
@@ -522,60 +514,88 @@ - (void)updateCardTitleBarBackgroundColors {
+        }
+        [card pop_addAnimation:springAnimation forKey:@"frame"];
+    }
+}
+
+- (void)updateCardLocationsWhileOpening {
+    if (self.currentCardIndex > 0) {
+        CGFloat startY = CardStackTopMargin;
+        CardView *currentCard = [self.cards objectAtIndex:self.currentCardIndex];
+        CGFloat endY = currentCard.frame.origin.y;
+        CGFloat currentY = startY;
+        CGFloat incrementY = (endY - startY) / self.currentCardIndex;
+        for (NSUInteger i = 0; i < self.currentCardIndex; i++) {
+            currentCard = [self.cards objectAtIndex:i];
+            CGRect frame = currentCard.frame;
+            frame.origin.y = currentY;
+            currentCard.frame = frame;
+            currentY = currentY + incrementY;
+        }
+    }
+}
+
+- (CGRect)frameForCardAtIndex:(NSUInteger)index {
+    CGRect frame;
+
+    BOOL isCardAtTheBottomAndCoveredByTheLastCard = (index > self.currentCardIndex && index < self.cards.count - 1);
+    if (index <= self.currentCardIndex) {
         if (self.isOpen) {
-            CGFloat previousTitleBarHeights = 0.0f;
-            for (NSUInteger i = 0; i < self.cards.count; i++) {
+            CGFloat previousTitleBarHeights = CardStackTopMargin;
+            for (NSUInteger i = 0; i < index; i++) {
                 CardView *card = [self.cards objectAtIndex:i];
 
-                CGRect frame = card.frame;
-                if (i <= self.currentCardIndex) {
-                    frame.origin.y = previousTitleBarHeights + CardStackTopMargin;
-                } else {
-                    frame.origin.y = self.view.bounds.size.height - card.titleBarHeight;
-                }
-                card.frame = frame;
-
                 // -1.0f is used to avoid area below the title bar to become slightly visible is some cases (due to rounding errors)
                 previousTitleBarHeights += (card.titleBarHeight * card.scale - 1.0f);
             }
-        } else {
-            for (NSUInteger i = 0; i < self.cards.count; i++) {
-                CardView *card = [self.cards objectAtIndex:i];
 
-                CGRect frame = card.frame;
-                if (i <= self.currentCardIndex) {
-                    frame.origin.y = 0;
-                } else {
-                    frame.origin.y = self.view.bounds.size.height - card.titleBarHeight;
-                }
-                card.frame = frame;
-            }
+            CardView *card = [self.cards objectAtIndex:index];
+            frame = card.frame;
+            frame.origin.y = previousTitleBarHeights;
+        } else {
+            CardView *card = [self.cards objectAtIndex:index];
+            frame = card.frame;
+            frame.origin.y = 0;
         }
+    } else if (isCardAtTheBottomAndCoveredByTheLastCard) {
+        // cards at the bottom but behind the last card will be positioned outside of the visible area, so their title bar won't show up when the last card is being moved
```

Comment:
I think it adds value: it explains _why_ certain cards are handled differently. In some cases I had trouble why I've done some exceptions even a few days later. Those are the point where I've added those comments.